### PR TITLE
Oracle client

### PR DESCRIPTION
**Pull request by y0sher**
server code is in https://github.com/y0sher/oracle_server for now.

This lets use a server as the authority of eligibility for testing. 
we use the json api in order to not pollute the node with server pb files

### CODE REVIEW

File: hare/algorithm.go [hunk 1]
```diff
@@ -213,7 +213,7 @@ func (proc *ConsensusProcess) validateRole(m *pb.HareMessage) bool {
 	// TODO: validate role proof
 
 	// validate role
-	if !proc.oracle.Validate(proc.expectedCommitteeSize(m.Message.K), Signature(m.Message.RoleProof)) {
+	if !proc.oracle.Validate(proc.instanceId.Bytes(), int(m.Message.K), proc.expectedCommitteeSize(m.Message.K), proc.signing.Verifier(), Signature(m.Message.RoleProof)) {
```
Comment (antonlerner) [inline hare/algorithm.go]: don't use bytes as id, set specific type

File: hare/algorithm.go [hunk 2]
```diff
@@ -514,7 +514,7 @@ func (proc *ConsensusProcess) endOfRound3() {
 }
 
 func (proc *ConsensusProcess) currentRole() Role {
-	if proc.oracle.Validate(proc.expectedCommitteeSize(proc.k), proc.roleProof()) {
+	if proc.oracle.Validate(proc.instanceId.Bytes(), int(proc.k), proc.expectedCommitteeSize(proc.k), proc.signing.Verifier(), proc.roleProof()) {
```
Comment (antonlerner) [inline hare/algorithm.go]: Is the verifier the public key and identity? if so, lets rename it to that

File: hare/mock_oracle.go [hunk 3]
```diff
@@ -96,7 +96,7 @@ func (mock *MockHashOracle) calcThreshold(committeeSize int) uint32 {
 }
 
 // Validate if a proof is valid for a given committee size
-func (mock *MockHashOracle) Validate(committeeSize int, proof Signature) bool {
+func (mock *MockHashOracle) Validate(instanceID []byte, K int, committeeSize int, pubKey Stringer, proof []byte) bool {
```
Comment (antonlerner) [inline hare/mock_oracle.go]: rename to validateEligibility

File: miner/block_builder.go [hunk 4]
```diff
@@ -159,17 +166,20 @@ func (t *BlockBuilder) acceptBlockData() {
...
-				blk := t.createBlock(id, txList)
-				go func() {
-					bytes, err := mesh.BlockAsBytes(blk)
-					if err != nil {
-						log.Error("cannot serialize block %v", err)
-						return
-					}
-					t.network.Broadcast(meshSync.NewBlock, bytes)
-				}()
+				if t.blockOracle.Validate(id, t.minerID) {
```
Comment (antonlerner) [inline miner/block_builder.go]: invert the if - 
if !Validate { 
break
}

File: oracle/oracle.go [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+func (s stringerer) String() string {
+	return string(s)
+}
+
+// ValidateBlock makes oracleBlock a BlockValidator
+func (bo *blockOracle) ValidateBlock(block mesh.Block) bool {
+	return bo.Validate(block.LayerIndex, stringerer(block.MinerID))
+}
+
+func (bo *blockOracle) Validate(id mesh.LayerID, pubKey fmt.Stringer) bool {
```
Comment (antonlerner) [inline oracle/oracle.go]: just accept string

File: oracle/oracle.go [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+func NewBlockOracle(worldid uint64, committeeSize int, pubKey fmt.Stringer) *blockOracle {
+	oc := NewOracleClientWithWorldID(worldid)
+	oc.Register(pubKey)
+	return &blockOracle{
+		committeeSize,
+		oc,
+	}
+}
+
+type stringerer string
```
Comment (antonlerner) [inline oracle/oracle.go]: use simple string as input

File: oracle/oracle.go [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+func (bo *blockOracle) Validate(id mesh.LayerID, pubKey fmt.Stringer) bool {
+	return bo.oc.Validate(common.Uint32ToBytes(uint32(id)), -1, bo.committeeSize, pubKey)
+}
+
+type HareOracle interface {
+	Validate(instanceID []byte, K int, committeeSize int, pubKey fmt.Stringer, proof []byte) bool
+}
+
+type hareOracle struct {
+	oc *OracleClient
```
Comment (antonlerner) [inline oracle/oracle.go:45]: you can embed OracleClient

File: oracle/rolacle_client.go [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+	c *http.Client
+}
+
+func NewHTTPRequester(url string) *HTTPRequester {
+	return &HTTPRequester{url, &http.Client{}}
+}
+
+func (hr *HTTPRequester) Get(api, data string) []byte {
+	var jsonStr = []byte(data)
+	spew.Println(string(jsonStr))
```
Comment (antonlerner) [inline oracle/rolacle_client.go]: you can just print data

File: oracle/rolacle_client.go [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+
+func hashInstanceAndK(instanceID []byte, K int) uint32 {
+	h := newHasherU32()
+	val := h.Hash(append(instanceID, byte(K)))
+	return val
+}
+
+
+// Validate checks whether a given ID is in the eligible list or not. it fetches the list once and gives answers locally after that.
+func (oc *OracleClient) Validate(instanceID []byte, K int, committeeSize int, pubKey fmt.Stringer) bool {
```
Comment (antonlerner) [inline oracle/rolacle_client.go]: ValidateElegibility

File: oracle/rolacle_client.go [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+	elgmap := make(map[string]struct{})
+
+	for _, v := range res.IDs {
+		elgmap[v] = struct{}{}
+	}
+
+	_, valid := elgmap[pubKey.String()]
+
+	oc.eligibilityMap[val] = elgmap
+	oc.instMtx[val].Unlock()
```
Comment (antonlerner) [inline oracle/rolacle_client.go]: acess to the map is done without the proper mutex locked (eMtx)

**Review by antonlerner — APPROVED** [review]

File: oracle/rolacle_client.go (hunk 9 above)
Comment (y0sher) [reply oracle/rolacle_client.go]: its `Validate` for following the hare Rolacle interface.

File: hare/mock_oracle.go (hunk 3 above)
Comment (y0sher) [reply hare/mock_oracle.go]: That was the original name I didn't change it. refactoring is out of this PR scope :)

**Pull request merged**